### PR TITLE
feat(dimensions): native per-page geometry without rendering

### LINKED ISSUE #33: dimensions: native per-page geometry (size + rotation) without rendering
## Summary

Add a native `pdq dimensions` command that reports each page's geometry (size in
PDF points + rotation) straight from the page tree — no rasterization — so
consumers can lay out a multi-page viewer **before** any page is rendered, and
handle mixed-size documents correctly.

## Motivation

The `info` command reports page dimensions by
rendering **page 1** at 72 dpi and reading the PNG size. That's a single-page
approximation: documents with mixed page sizes are reported wrong for every page
but the first.

Real example: a 48-page menu whose pages are `419×595` pt except the last, which
is `340×680` pt. A viewer trusting `/info` places page 48 at the wrong size.

The viewer wants **all** page sizes up front (scrollbar height, virtualized page
positions, placeholder boxes) before it fetches any page image. Rendering every
page just to measure it is wasteful (and slow on large docs); the geometry is
already in each page's `MediaBox`/`CropBox` + `/Rotate`, which pdq reads during
`page-count`/`render` anyway.

## Proposed CLI

```
pdq dimensions [--password <PW>] <INPUT>
```

- Output JSON on stdout. Cheap: page-tree walk only, no content-stream interpret.

## Proposed output (points at 72 dpi)

```json
{
  "pages": 48,
  "page_sizes": [
    { "width": 419, "height": 595 },
    { "width": 340, "height": 680 }
  ]
}
```

- `width`/`height` in **PDF points**, with `/Rotate` applied (a page rotated
  90/270 reports swapped width/height), so the numbers match what `pdq render`
  produces at scale 1 — i.e. `render` px = `width * dpi/72`.
- Effective box = CropBox if present, else MediaBox (match `render`).

## Requirements / edge cases

- Rotation normalized to the visible orientation (swap on 90/270), consistent
  with `render`.
- Encrypted inputs via `--password` (like `page-count`).
- Damaged/`/Rotate`-missing pages: fall back to MediaBox / 0 rotation rather than
  failing the whole document.

## Acceptance

- For every page, `width`/`height` × (dpi/72) equals the pixel size of
  `pdq render` of that page at that dpi.
- Mixed-size documents report each page's true size (not page 1's).

### PR DESCRIPTION
**Pull request by Guiqft**
## Summary

Adds a native `pdq dimensions` command (and `pdq::page_dimensions` /
`page_dimensions_with_password` library API) that reports every page's size in
PDF points and its rotation straight from the page tree — no rasterization:

```
$ pdq dimensions [--password PW] input.pdf
{"pages":48,"page_sizes":[{"width":419,"height":595,"rotation":0},...,{"width":340,"height":680,"rotation":0}]}
```

Mixed-size documents report each page's true size, so a viewer can lay out
scrollbar height, virtualized positions, and placeholder boxes before fetching
any page image.

Closes #33

## How it works

- Rides on the same lazy, mmap-backed reader as `page-count`: the page list
  comes from the shared page-tree walk (it can never disagree with the pages
  `split`/`render` resolve), `--password` and transparent owner-password
  decryption work as elsewhere, and damaged xrefs get the usual repair retry.
- `MediaBox`/`CropBox`/`Rotate` are resolved through page-tree inheritance by
  walking `/Parent` chains with a per-node cache, tolerating indirect
  references at both the array and coordinate level.
- The geometry math replicates hayro's `Page::render_dimensions` exactly —
  effective box = CropBox∩MediaBox (CropBox defaults to MediaBox), missing
  MediaBox → A4, nearly-zero box → A4, per-dimension `max(1.0)`, the same
  f64→f32 cast points, `rem_euclid(360)` rotation normalization with 90/270
  swapping width/height — so `floor(width × dpi/72)`, computed in f32 like hayro computes it,
  equals the pixel size `pdq render` produces at that dpi. The precision
  matters: in f64 the formula predicts 1px too many on some DPI/page-size
  combinations (a Letter page at 150 dpi renders 1275×1649, not the 1275×1650
  doubles suggest), so the README documents the exact single-precision recipe
  for viewer authors (`Math.fround` in JS). A parity test asserts prediction ==
  rendered PNG size at 72, 96, 144, 150, 200, and 300 dpi, pinning the 150-dpi
  Letter case where f32 and f64 disagree.

## Edge cases

- Per-page damage degrades instead of failing the document: a missing or
  malformed box falls back to the inherited value, then A4; a malformed
  `/Rotate` falls back to 0.
- A box array with a non-numeric element is rejected whole (matching hayro's
  stop-at-first-non-numeric parsing) rather than shifting later coordinates
  into the wrong slots.
- Coordinates that overflow f32 to infinity fall back to A4: `render` refuses
  such pages ("too large to render"), and bare `inf` is not valid JSON.
- `rotation` is included per entry (the issue title asks for size + rotation),
  normalized to 0/90/180/270.

The command works without the `render` feature — it is a pure metadata walk.

### CODE REVIEW

File: src/dimensions.rs [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+    let x0 = coords.next()?? as f64;
+    let y0 = coords.next()?? as f64;
+    let x1 = coords.next()?? as f64;
+    let y1 = coords.next()?? as f64;
+    Some(Rect {
+        x0: x0.min(x1),
+        y0: y0.min(y1),
+        x1: x1.max(x0),
+        y1: y1.max(y0),
+    })
```
[the remark this answers is not in the document]
Comment (Guiqft) [reply src/dimensions.rs:262]: False positive: struct-literal field expressions evaluate against the enclosing scope's locals — `x0: x0.min(x1)` does not create a binding that later fields see, so `x1: x1.max(x0)` still reads the original local `x0`. Verified empirically with `/MediaBox [600 0 0 800]`: `pdq dimensions` reports `600×800` and `pdq render` at 72 dpi produces a 600×800 px PNG, so the corner normalization works and parity with `render` holds. (This is also byte-for-byte hayro's own `from_arr` expression — `hayro-syntax/src/object/rect.rs` — which `render` relies on.)

File: src/dimensions.rs [hunk 2]
```diff
@@ -0,0 +1,292 @@
...
+}
+
+fn dimensions_impl(source: &PdfSource) -> Result<Vec<PageDimensions>> {
+    let page_ids = source.page_ids()?;
+    let mut nodes = NodeCache::default();
+    Ok(page_ids
+        .into_iter()
+        .map(|id| resolve_page(source, &mut nodes, id))
+        .collect())
+}
```
[the remark this answers is not in the document]
Comment (Guiqft) [reply src/dimensions.rs:99]: Consistent by design: this mirrors `page-count`, which prints `0` with a success exit for the same document — documented in `count.rs` ("Returns `0` for a structurally valid PDF that declares no pages"). The paths that reject an empty page tree (`split`/`merge` via `PageRangeError::NoPages`) do so because they must produce output documents; the read-only reporting commands both treat zero pages as an accurate answer. For the viewer-layout use case in #33, `{"pages":0,"page_sizes":[]}` is the honest result to lay out.

**Comment by Guiqft** [comment]
@greptile-apps review again
